packages/controller/src/lib/adapterAutoUpgradeManager.ts

```diff
@@ -0,0 +1,195 @@
+                    upgradedAdapters.push({
+                        name: repoAdapterInfo.name,
+                        newVersion: repoAdapterInfo.version,
+                        oldVersion: adapterConfig.version
+                    });
+                    this.logger.info(
+                        `${this.logPrefix} Successfully upgraded adapter "${repoAdapterInfo.name}" to ${repoAdapterInfo.version}`
+                    );
+                } catch (e) {
+                    this.logger.error(
+                        `${this.logPrefix} Could not upgrade adapter "${repoAdapterInfo.name}" to ${repoAdapterInfo.version}: ${e.message}`
+                    );
+                }
+            }
+        }
+
+        return upgradedAdapters;
+    }
+
+    /**
+     * Upgrade specified adapter to given version
+     *
+     * @param options information of the adapter to install, e.g. version and name, sa well as active repo
+     */
+    private async upgradeAdapter(options: UpgradeAdapterOptions): Promise<void> {
+        const { repoName, name, version } = options;
+
+        this.logger.info(`${this.logPrefix} Upgrade adapter "${name}" to ${version}`);
+
+        const upgrade = new Upgrade({
+            objects: this.objects,
+            states: this.states,
+            params: {},
+            processExit: () => undefined
+        });
+
+        await upgrade.upgradeAdapter(repoName, `${name}@${version}`, false, true, false);
+    }
+
+    /**
+     * Get the current active repository name
+     */
+    private async getConfiguredRepositoryName(): Promise<string> {
+        const obj = await this.objects.getObjectAsync('system.config');
+
+        if (!obj?.common?.activeRepo?.length) {
+            throw new Error('Could not find an active repository');
+        }
+
+        return obj.common.activeRepo[0];
+    }
+
+    /**
+     * Get the repository information for the given repository
+     *
+     * @param name Name of the repository
+     */
+    private async getRepository(name: string): Promise<Record<string, RepositoryAdapter>> {
+        const obj = await this.objects.getObjectAsync('system.repositories');
+
+        if (!obj?.native?.repositories?.[name]) {
+            throw new Error(`Could not get repository information for "${name}"`);
+        }
+
+        delete obj.native.repositories[name]._repoInfo;
+        return obj.native.repositories[name].json;
+    }
+
+    /**
+     * Get the auto upgrade configuration of all adapters
+     */
+    private async getAutoUpdateConfiguration(): Promise<AdapterUpgradeConfiguration[]> {
+        const res = await this.objects.getObjectViewAsync('system', 'adapter', {
+            startkey: 'system.adapter.',
+            endkey: 'system.adapter.\u9999'
+        });
+
+        if (!res) {
+            throw new Error('Did not get information about installed adapters');
+        }
+
+        return res.rows
+            .filter(row => row.value.common.automaticUpgrade !== 'none')
```

Comment:
how to make sure it is set ? or also exclude undefined?

Reply:
interesting that it has not crashed during my tests